### ISSUE TITLE
`inject_hyperparams` causes numerical divergence in Adam compared to standard optimizer with identical schedule

### ISSUE THREAD

**Issue by kameronton** [google-deepmind/optax#1532]
I noticed that wrapping an optimizer with `optax.inject_hyperparams` results in optimization trajectories that diverge from the unwrapped optimizer, even when both are initialized with the exact same constant schedule and data. Specifically, there is a difference in both loss trajectories and learned parameters of the network. While minor, it causes a noticeable deviation in my application case (RL), especially long-term.

I checked different optimizers:
- divergence observed: `optax.adam`, `optax.amsgrad`
- no divergence: `optax.sgd`, `optax.adagrad`

What I would like to know is if this behavior is intended and if there is a way to force alignment. Also might be related to #583 .

An MWE uses `optax.adam` and a simple feed-forward NN that fits a randomly generated dataset:

```py
import jax
import jax.numpy as jnp
import optax
import matplotlib.pyplot as plt
from flax import linen as nn

# an MLP to optimize
class MLP(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.Dense(32)(x)
        x = nn.relu(x)
        x = nn.Dense(32)(x)
        x = nn.relu(x)
        x = nn.Dense(1)(x)
        return x

# synthetic data
key = jax.random.PRNGKey(0)
k1, k2 = jax.random.split(key)
X = jax.random.uniform(k1, (100, 1), minval=-3, maxval=3)
Y = jnp.sin(X) + jax.random.normal(k2, (100, 1)) * 0.1


model = MLP()
init_params = model.init(jax.random.PRNGKey(42), jnp.ones((1, 1)))


def loss_fn(params, x, y):
    pred = model.apply(params, x)
    return jnp.mean((pred - y) ** 2)

epochs = 500
schedule = optax.constant_schedule(jnp.array(1e-3))

# the usual initialization
tx_standard = optax.adam(learning_rate=schedule)
state_standard = tx_standard.init(init_params)
params_standard = init_params

@jax.jit
def step_standard(params, opt_state, x, y):
    grads = jax.grad(loss_fn)(params, x, y)
    updates, new_opt_state = tx_standard.update(grads, opt_state, params)
    new_params = optax.apply_updates(params, updates)
    loss = loss_fn(new_params, x, y)
    return new_params, new_opt_state, loss

# lr using injected hyperparams
tx_injected = optax.inject_hyperparams(optax.adam)(learning_rate=schedule)
state_injected = tx_injected.init(init_params)
params_injected = init_params

@jax.jit
def step_injected(params, opt_state, x, y):
    grads = jax.grad(loss_fn)(params, x, y)
    updates, new_opt_state = tx_injected.update(grads, opt_state, params)
    new_params = optax.apply_updates(params, updates)
    loss = loss_fn(new_params, x, y)
    return new_params, new_opt_state, loss


losses_std = []
losses_inj = []
diffs = []
losses_diff = []

for i in range(epochs):
    params_standard, state_standard, loss_std_val = step_standard(
        params_standard, state_standard, X, Y
    )
    losses_std.append(loss_std_val)

    params_injected, state_injected, loss_inj_val = step_injected(
        params_injected, state_injected, X, Y
    )
    losses_inj.append(loss_inj_val)

    # track difference in losses
    losses_diff.append(jnp.abs(loss_std_val - loss_inj_val))

    # calculate distance between parameters of both models to see if they diverge
    param_diff = sum(
        jnp.sum((p1 - p2)**2) 
        for p1, p2 in zip(jax.tree_util.tree_leaves(params_standard), 
                          jax.tree_util.tree_leaves(params_injected))
    )
    diffs.append(param_diff)

# plotting
fig, axes = plt.subplots(3, figsize=(15, 8))

# loss curves
axes[0].plot(losses_std, label='Standard', linewidth=4, alpha=0.5)
axes[0].plot(losses_inj, label='Injected', linestyle='--', color='red')
axes[0].set_title('Loss Function Trajectory')
axes[0].set_xlabel('Epoch')
axes[0].set_ylabel('MSE Loss')
axes[0].legend()
axes[0].grid(True, alpha=0.3)

# loss difference
axes[1].plot(losses_diff)
axes[1].set_title('Loss Difference Between Methods')
axes[1].set_xlabel('Epoch')
axes[1].set_ylabel('|Loss_1 - Loss_2|')
axes[1].grid(True, alpha=0.3)

# parameter divergence
axes[2].plot(diffs)
axes[2].set_yscale('log')
axes[2].set_title('Parameter Divergence (L2 Distance)')
axes[2].set_xlabel('Epoch')
axes[2].set_ylabel('||Params_1 - Params_2||^2')
axes[2].grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

```

<img width="1500" height="800" alt="Image" src="https://github.com/user-attachments/assets/4d0ffb2a-0929-4ab0-abed-5a350677857c" />

**Comment by kameronton** [comment]
I've looked a bit more at it and was able to identify the cause. In particular, the difference can be seen in the internal states of the optimizer after the very first step, which indicates that some precision/type conversion processes are involved, see `mu` and `nu` (I used a simpler function to optimize here, something like `jnp.sum(params ** 2)` for two parameters already shows the difference):

```
opt_state_std = (ScaleByAdamState(count=Array(1, dtype=int32), mu=Array([ 1., -1.], dtype=float32), nu=Array([0.1, 0.1], dtype=float32)), ScaleByScheduleState(count=Array(1, dtype=int32)))
opt_state_inj = InjectStatefulHyperparamsState(count=Array(1, dtype=int32), hyperparams={'b1': Array(0.9, dtype=float32), 'b2': Array(0.999, dtype=float32), 'eps': Array(1.e-08, dtype=float32), 'eps_root': Array(0., dtype=float32), 'learning_rate': Array(0.1, dtype=float32)}, hyperparams_states={'learning_rate': WrappedScheduleState(count=Array(1, dtype=int32))}, inner_state=(ScaleByAdamState(count=Array(1, dtype=int32), mu=Array([ 1.0000002, -1.0000002], dtype=float32), nu=Array([0.09999871, 0.09999871], dtype=float32)), EmptyState()))
```

`optax.inject_hyperparams` recreates the inner transformation with its hyperparameters converted to the `dtype` that is inferred from the params/updates:
```py
      hparams = {
          k: _convert_floats(v, dtype) for k, v in state.hyperparams.items()
      }
      hparams.update({
          k: _convert_floats(
              f(state.hyperparams_states[k], **extra_args), dtype
          )
          for k, f in sched_hps.items()
      })
```

Literal values used in `adam` like `b1=0.9` are immediately turned into `float32` when the params are `float32`, so the coefficients that scale the moments already contain rounding error even before the update. The same happens for Amsgrad. 

For now, enabling double precision and setting `hyperparam_dtype=jnp.float64` when calling `inject_hyperparams` is a workaround.

**Comment by kameronton** [comment]
I checked two potential fixes for this issue, but both involve trade-offs regarding the current API. I would appreciate guidance on which direction is preferable before submitting a PR.

**Option 1**: Preserve Python literals

The idea is to modify `inject_hyperparams` to avoid casting Python scalars to `dtype` unless `hyperparam_dtype` is explicitly provided. This allows parameters to remain Python literals during tracing, matching the behavior of the standard optimizer.

+: Solves the divergence immediately by preserving literals.

-: These hyperparameters become "baked in" to the XLA graph as constants. If a user later attempts to mutate `state.hyperparams['b1']`, the change will be ignored. Users are required to explicitly define a `dtype` during initialization if they intend to dynamically mutate hyperparameters later.

A less invasive option would be to have a flag `keep_scalars` in `inject_hyperparams` that would turn off casting to `dtype` and force the behavior to be aligned. In this case, however, the default behavior can be considered as faulty.

**Option 2**: Only store explicitly injected params

The idea is to change the design so that `inject_hyperparams` only captures and stores parameters that are actually passed to the wrapper (e.g., just `learning_rate`), rather than inspecting the inner optimizer's signature and capturing all defaults (like `b1`, `b2`, `eps`) into the state and changing their type.

+: A cleaner solution: constants remain constants in the closure, and only dynamic hyperparameters are treated as state (with the type casting).

-: It appears the current design intent of `inject_hyperparams` is to expose all arguments for introspection. Restricting the state to only explicitly injected parameters causes many test failures, since many tests expect full visibility of all optimizer arguments in state.hyperparams.

**Comment by rdyro** [comment]
Hey, thanks for investigating this. Option 2 is an interesting idea and sounds much better than Option, but I don't know if either is feasible.

A numerics change like that is generally something I'd have to author on our side since it can affect numerics.

I definitely sympathize with the unexpected numerics change in your first comment, but it seems that despite the divergence, the loss doesn't differ by much more than numerical precision. It might be necessary to keep the behavior as is unless you run into a divergence of much higher magnitude.

**Comment by kameronton** [comment]
@rdyro , thanks for the reply! I completely understand the caution regarding changing numerics. I'd like to highlight two points.

1. To address the point about numerical precision: simply take the example from my initial post and increase the number of layers from two to four. The difference in losses gets as big as 3e-4 (see picture). If you consider deeper architectures, a problem with less smooth loss landscape, or run the training for longer, then the differences become way more noticeable. The resulting models converge to different minima.

<img width="1500" height="800" alt="Image" src="https://github.com/user-attachments/assets/df161d55-d019-4f21-926c-7105c1a58207" />

2. The main concern is reproducibility. My understanding is that one of the ways people use `inject_hyperparams` is to simply track the hyperparams during training, without necessarily changing them. When used in this mode, current implementation changes the training trajectories, breaking transparency. I guess if not fixed, this side effect that should be stated somewhere in the documentation.

Regarding feasibility, I made a proof of concept of the less invasive version of option 1, see https://github.com/kameronton/optax/commit/e982277d1b4ce237a1d339057b8c965a9e7dc572 . This implementation keeps scalars as literals if a flag is given to `inject_hyperparams`.

Let me know what you think!

**Comment by rdyro** [comment]
Thanks for investigating this further.

The one confusing part is that when passing state through a jitted function, the float hyperparameters will get converted to an array anyway. It's possible to avoid that if you never pass the state out or into jit (e.g., several steps under the same jit), but it might get confusing.

Somewhat unsatisfyingly, you can still keep the e.g., learning_rate rate as a float while treating other parameters as hyperparameters if you specify `optax.inject_hyperparams(optax.adam, static_args=("learning_rate",))(1e-3)`, but you can't monitor the hyperparameter value then (it's not part of the hyperparam-state).

I'll close the issue, since I'm not sure your solution gets us quite there in general (because of the jit issue), but feel free to re-open if you want to continue the discussion.
